### PR TITLE
feat(get_transaction_count): add

### LINKED ISSUE #101: feat: eth_getTransactionCount
# eth_getTransactionCount

- name: getTransactionCount
- prefix: eth
- state: ⚠️
- [specification](https://github.com/ethereum/execution-apis)

## Specification Description

Requests a specific account's number of transactions, in order to compute its nonce.

### Parameters

- Address
- Block

### Returns

- uint256 - e.g. Transaction Count

## Kakarot Logic

Call the Starknet `starknet_getNonce` json rpc method. Return it.

### PR DESCRIPTION
**Pull request by internnos**
Time spent on this PR:
1 day

Closes #101 

# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing
- [ ] Other (please describe):

# What is the current behavior?

getTransactionCount is hardcoded to 3

Issue Number: N/A

# What is the new behavior?

-
-
-

# Does this introduce a breaking change?

- [ ] Yes
- [ ] No

# Other information

### CODE REVIEW

File: core/src/client/client_api.rs [hunk 1]
```diff
@@ -121,6 +121,12 @@ pub trait KakarotClient: Send + Sync {
         block: MaybePendingStarknetBlock,
     ) -> Result<RichBlock, KakarotClientError>;
 
+    async fn get_transaction_count(
+        &self,
+        ethereum_address: Address,
+        _block_number: Option<BlockId>,
```
Comment (greged93) [inline core/src/client/client_api.rs:127]: Why the `_` for block_number?

File: rpc/src/eth_api.rs [hunk 2]
```diff
@@ -119,8 +119,8 @@ pub trait EthApi {
     #[method(name = "eth_getTransactionCount")]
     async fn transaction_count(
         &self,
-        address: Address,
```
Comment (greged93) [inline rpc/src/eth_api.rs:122]: Why did you add the `_` for address and block_number? In rust, you prefix your variable name in order to inform the compiler that you are aware of the unused variable. But in our case address and block_number are used.

**Review by greged93 — COMMENTED** [review]
hey @internnos, thank you for your PR! Just some comments below

**Review by greged93 — CHANGES_REQUESTED** [review]
Can you also add a test in `rpc/tests/rpc.rs` ?

File: rpc/src/eth_api.rs (hunk 2 above)
Comment (internnos) [reply rpc/src/eth_api.rs:122]: oh i see, gotcha will remove it

File: core/src/client/client_api.rs (hunk 1 above)
Comment (internnos) [reply core/src/client/client_api.rs:127]: will remove it, on it!

**Comment by internnos** [comment]
> get_transaction_count

oh yeah ofc! let me try

**Comment by internnos** [comment]
> Can you also add a test in `rpc/tests/rpc.rs` ?

hey @greged93 do you have any idea on how to inject some values / mock some object? the thing is, I always got `
"Kakarot Core: Light Client Request Error: Contract not found"` (expected) when testing. the other option is to run starknet devnet, and run the test, since the devnet already have some contracts

**Comment by greged93** [comment]
Hey @internnos, in `rpc/tests/rpc.rs` we are testing using the `wiremock` crate. [Wiremock](https://crates.io/crates/wiremock) allows you to mimic a HTTP server. You can define the expected requests and responses in `core/src/utils/wiremock_utils.rs`. You should:
- Define a test for `get_transaction_count` in `rpc/tests/rpc.rs`.
- Define the expected request and response in `core/src/utils/wiremock_utils.rs` for testing `get_transaction_count` and mount it on the MockServer in `setup_wiremock`.

Let me know if you have more questions

**Comment by internnos** [comment]
> Hey @internnos, in `rpc/tests/rpc.rs` we are testing using the `wiremock` crate. [Wiremock](https://crates.io/crates/wiremock) allows you to mimic a HTTP server. You can define the expected requests and responses in `core/src/utils/wiremock_utils.rs`. You should:
> 
> * Define a test for `get_transaction_count` in `rpc/tests/rpc.rs`.
> * Define the expected request and response in `core/src/utils/wiremock_utils.rs` for testing `get_transaction_count` and mount it on the MockServer in `setup_wiremock`.
> 
> Let me know if you have more questions

hey @greged93, gotcha, will do it! thanks for the help!

**Comment by internnos** [comment]
> 

hi @greged93 i have a question

the thing is, the method i'd like to mock is called `compute_starknet_address` and its not a http request.
after i can compute the starknet address, then i'll call the `starknet_getNonce` endpoint(mocked using your suggestion)

**Comment by greged93** [comment]
`compute_starknet_address` is an entrypoint of the kakarot contract. In order to mock a call to a starknet contract, you can have a look at what was done for `mock_get_evm_address` in the wiremock. This should help you in order to implement `mock_compute_starknet_address`.

**Comment by internnos** [comment]
> `compute_starknet_address` is an entrypoint of the kakarot contract. In order to mock a call to a starknet contract, you can have a look at what was done for `mock_get_evm_address` in the wiremock. This should help you in order to implement `mock_compute_starknet_address`.

ahh gotcha! thanks!

**Comment by Eikix** [comment]
hi:), what is the status on this?

**Comment by Eikix** [comment]
@bajpai244 will now take this issue

**Comment by bajpai244** [comment]
Yes, I have made a new issue #135 for this as well!

**Comment by Eikix** [comment]
Closing, as we are in refactor cycle and PR has gone stale
